// @markuplint/rules/lib/permitted-contents/recursive-branch.js
import { matchesSelector } from './matches-selector.js';
import { order } from './order.js';
import { Collection, isModel } from './utils.js';
export function recursiveBranch(model, 
// eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
childNodes, specs, options, depth) {
    if (!isModel(model)) {
        return order(model, childNodes, specs, options, depth + 1);
    }
    if (typeof model === 'string') {
        return matchesSelector(model, childNodes[0], specs, depth);
    }
    const collection = new Collection(childNodes);
    let lastUnmatched = null;
    for (const query of model) {
        const result = matchesSelector(query, collection.unmatched[0], specs, depth);
        collection.addMatched(result.matched);
        if (result.type === 'MATCHED' || result.type === 'MATCHED_ZERO') {
            return {
                type: result.type,
                matched: collection.matched,
                unmatched: collection.unmatched,
                zeroMatch: result.zeroMatch,
                query: result.query,
                hint: result.hint,
            };
        }
        lastUnmatched = {
            type: result.type,
            matched: collection.matched,
            unmatched: collection.unmatched,
            zeroMatch: result.zeroMatch,
            query: result.query,
            hint: result.hint,
        };
    }
    if (!lastUnmatched) {
        throw new Error('Unreachable code');
    }
    return lastUnmatched;
}
