// @markuplint/rules/lib/table-row-column-alignment/grid.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _Grid_tbodyRowElements, _Grid_tfootRowElements, _Grid_theadRowElements;
import { findChildren } from './find-children.js';
export class Grid {
    constructor(
    // eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
    table) {
        _Grid_tbodyRowElements.set(this, void 0);
        _Grid_tfootRowElements.set(this, void 0);
        _Grid_theadRowElements.set(this, void 0);
        const thead = findChildren(table, 'thead')[0];
        const tbody = findChildren(table, 'tbody')[0];
        const tfoot = findChildren(table, 'tfoot')[0];
        __classPrivateFieldSet(this, _Grid_theadRowElements, thead ? findChildren(thead, 'tr') : [], "f");
        __classPrivateFieldSet(this, _Grid_tbodyRowElements, tbody ? findChildren(tbody, 'tr') : [], "f");
        __classPrivateFieldSet(this, _Grid_tfootRowElements, tfoot ? findChildren(tfoot, 'tr') : [], "f");
        this.theadGrid = createGrid(__classPrivateFieldGet(this, _Grid_theadRowElements, "f"));
        this.tbodyGrid = createGrid(__classPrivateFieldGet(this, _Grid_tbodyRowElements, "f"));
        this.tfootGrid = createGrid(__classPrivateFieldGet(this, _Grid_tfootRowElements, "f"));
    }
    getAllRowElements() {
        return [...__classPrivateFieldGet(this, _Grid_theadRowElements, "f"), ...__classPrivateFieldGet(this, _Grid_tbodyRowElements, "f"), ...__classPrivateFieldGet(this, _Grid_tfootRowElements, "f")];
    }
    getAllRows() {
        return [
            ...this.theadGrid.filter(hasElementFilter),
            ...this.tbodyGrid.filter(hasElementFilter),
            ...this.tfootGrid.filter(hasElementFilter),
        ];
    }
    getBaseColLength() {
        if (this.theadGrid.length > 0) {
            return getBaseColLength(this.theadGrid);
        }
        if (this.tfootGrid.length > 0) {
            return getBaseColLength(this.tfootGrid);
        }
        return getBaseColLength(this.tbodyGrid);
    }
    getSections() {
        return [
            {
                section: this.theadGrid,
                elements: __classPrivateFieldGet(this, _Grid_theadRowElements, "f"),
            },
            {
                section: this.tbodyGrid,
                elements: __classPrivateFieldGet(this, _Grid_tbodyRowElements, "f"),
            },
            {
                section: this.tfootGrid,
                elements: __classPrivateFieldGet(this, _Grid_tfootRowElements, "f"),
            },
        ];
    }
    hasOverlapped() {
        return (this.tbodyGrid.some(row => row.includes('x')) ||
            this.tfootGrid.some(row => row.includes('x')) ||
            this.theadGrid.some(row => row.includes('x')));
    }
    log() {
        // eslint-disable-next-line no-console
        console.table(this.theadGrid);
        // eslint-disable-next-line no-console
        console.table(this.tbodyGrid);
        // eslint-disable-next-line no-console
        console.table(this.tfootGrid);
    }
}
_Grid_tbodyRowElements = new WeakMap(), _Grid_tfootRowElements = new WeakMap(), _Grid_theadRowElements = new WeakMap();
export function getOverflowRowSpan(rows, 
// eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
rowElements) {
    for (const [rowNum] of rows.entries()) {
        const rowEl = rowElements[rowNum];
        if (!rowEl) {
            const prevRow = rows[rowNum - 1];
            const prevRowEl = rowElements[rowNum - 1];
            if (!prevRow || !prevRowEl) {
                continue;
            }
            const spanStart = prevRow.indexOf('◎');
            if (spanStart === -1) {
                continue;
            }
            const indexes = getIndexes(prevRow);
            const index = indexes[spanStart];
            if (index == null) {
                continue;
            }
            const cells = findChildren(prevRowEl, 'th, td');
            const spanStartCell = cells[index];
            if (!spanStartCell) {
                continue;
            }
            const rowSpan = spanStartCell.getAttributeNode('rowspan');
            if (!rowSpan) {
                continue;
            }
            return {
                rowSpan,
            };
        }
    }
    return null;
}
export function getIndexes(row) {
    let indexCounter = 0;
    return row.map(col => (col === '●' || col === '◎' ? indexCounter++ : null));
}
function getBaseColLength(grid) {
    let baseColLength;
    if (grid.length >= 3) {
        const totalCols = grid.reduce((acc, row) => {
            return acc + row.length;
        }, 0);
        const average = Math.round(totalCols / grid.length);
        baseColLength = grid
            // eslint-disable-next-line unicorn/no-array-reduce
            .reduce((closest, row) => {
            return Math.abs(row.length - average) < Math.abs(closest - average) ? row.length : closest;
        }, grid[0]?.length ?? 0);
    }
    else {
        baseColLength = grid[0]?.length ?? 0;
    }
    return baseColLength;
}
function createGrid(
// eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
rows) {
    const rowSpans = new Set();
    const matrix = [];
    for (const [rowNum, row] of rows.entries()) {
        const cols = findChildren(row, 'th, td');
        let colNum = 0;
        matrix[rowNum] = matrix[rowNum] ?? [];
        for (const col of cols) {
            const colSpan = Number.parseInt(col.getAttribute('colspan') ?? '1');
            const rowSpan = Number.parseInt(col.getAttribute('rowspan') ?? '1');
            for (let i = 0; i < colSpan; i++) {
                const key = `${rowNum}:${colNum}`;
                let spanStart = false;
                if (colSpan > 1 && i === 0) {
                    spanStart = true;
                }
                if (rowSpan > 1) {
                    if (i === 0) {
                        spanStart = true;
                    }
                    for (let i = 1; i <= rowSpan - 1; i++) {
                        matrix[rowNum + i] = matrix[rowNum + i] ?? [];
                        matrix[rowNum + i][colNum] = '↓';
                        rowSpans.add(`${rowNum + i}:${colNum}`);
                    }
                }
                const addCol = rowSpans.has(key) ? 1 : 0;
                if (matrix[rowNum]?.[colNum] === '↓' && !spanStart && colSpan > 1) {
                    // Overlap
                    matrix[rowNum][colNum] = 'x';
                }
                else {
                    matrix[rowNum][colNum] = spanStart ? '◎' : addCol > 0 ? '↓' : colSpan > 1 ? '→' : '●';
                }
                if (addCol === 1) {
                    matrix[rowNum].push('●');
                }
                colNum += 1;
            }
        }
    }
    return matrix;
}
function hasElementFilter(row) {
    return row.some(cell => cell === '●' || cell === '◎');
}
