// @markuplint/ml-core/lib/ml-dom/node/block.js
import { after, before, remove, replaceWith } from '../manipulations/child-node-methods.js';
import { MLNode } from './node.js';
export class MLBlock extends MLNode {
    constructor(astNode, 
    // eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
    document) {
        super(astNode, document, astNode.isFragment);
        // TODO:
        this.isTransparent = true;
        this.conditionalType = astNode.conditionalType;
    }
    /**
     * Returns a string appropriate for the type of node as `MLBlock`
     *
     * @implements `@markuplint/ml-core` API: `MLBlock`
     */
    get nodeName() {
        return '#ml-block';
    }
    /**
     * Returns a number appropriate for the type of `MLBlock`
     *
     * @implements `@markuplint/ml-core` API: `MLBlock`
     */
    get nodeType() {
        return this.MARKUPLINT_PREPROCESSOR_BLOCK;
    }
    /**
     * @implements DOM API: `ChildNode`
     */
    after(
    // eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
    ...nodes) {
        after(this, ...nodes);
    }
    /**
     * @implements DOM API: `ChildNode`
     */
    before(
    // eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
    ...nodes) {
        before(this, ...nodes);
    }
    /**
     * @implements DOM API: `ChildNode`
     */
    remove() {
        remove(this);
    }
    /**
     * @implements DOM API: `ChildNode`
     */
    replaceWith(
    // eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
    ...nodes) {
        replaceWith(this, ...nodes);
    }
}
